### PR TITLE
Add 3.12 support

### PR DESCRIPTION
**Pull request by albi3ro**
This PR adds Python 3.12 support.

### CODE REVIEW

File: doc/releases/changelog-0.34.0.md [hunk 1]
```diff
@@ -441,6 +441,8 @@
 
 <h4>Other improvements</h4>
 
+* PennyLane now supports Python 3.12.
```
Comment (albi3ro) [inline doc/releases/changelog-0.34.0.md:444]:
```suggestion
* PennyLane now supports Python 3.12.
  [(#4985)](https://github.com/PennyLaneAI/pennylane/pull/4985)
```

**Comment by albi3ro** [comment]
[sc-53013]

**Review by mudit2812 — COMMENTED** [review]
Seems like the exact error raised for poorly defined children of abstract base classes has changed in 3.12. The regex used to validate that the correct error is raised in the failing test needs to be shortened, but otherwise looks good. Happy to approve once resolved.

File: .github/workflows/interface-unit-tests.yml [hunk 2]
```diff
@@ -70,10 +70,10 @@ jobs:
           else
             cat >python_versions.json <<-EOF
           {
-            "default": ["3.9", "3.10", "3.11"],
+            "default": ["3.9", "3.10", "3.11", "3.12"],
             "torch-tests": ["3.9", "3.11"],
-            "tf-tests": ["3.9", "3.10"],
-            "jax-tests": ["3.9", "3.11"],
+            "tf-tests": ["3.9", "3.11"],
```
Comment (timmysilv) [inline .github/workflows/interface-unit-tests.yml:75]: 😅

**Review by timmysilv — APPROVED** [review]

**Review by mudit2812 — APPROVED** [review]

**Pull request merged**